string-template/src/main/java/org/jdbi/v3/stringtemplate/StringTemplateSqlLocator.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jdbi.v3.stringtemplate;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.nio.charset.Charset;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+import org.antlr.stringtemplate.StringTemplateGroup;
+
+/**
+ * Locates SQL in <code>.sql.stg</code> StringTemplate files on the classpath.
+ */
+public class StringTemplateSqlLocator {
+    private static final ConcurrentMap<Class<?>, StringTemplateGroup> CACHE = new ConcurrentHashMap<>();
+    private static final Charset UTF_8 = Charset.forName("UTF-8");
+    private static final String TEMPLATE_GROUP_EXTENSION = ".sql.stg";
+
+    private StringTemplateSqlLocator() {
+    }
+
+    public static String findStringTemplateSql(Class<?> sqlObjectType, String name) {
+        StringTemplateGroup group = CACHE.computeIfAbsent(sqlObjectType, StringTemplateSqlLocator::readGroupResource);
+
+        if (!group.isDefined(name)) {
+            throw new IllegalStateException("No StringTemplate group " + name + " for class " + sqlObjectType);
+        }
+        return group.getInstanceOf(name).getTemplate();
```

Comment:
This `getTemplate()` call is the only thing I have reservations about. I'm uncertain whether references between templates in a group will resolve:

    group Foo;
    select() ::= << select * from foo >>
    selectById() ::= << $select() where id = :id >>